// @pulse.io/components/dist/collection/components/pulse-atm/checkbox/pulse-checkbox.js
import { h } from '@stencil/core';
import { createColorClasses } from '../../../utils/themes';
export class PulseCheckBox {
    constructor() {
        this.disabled = false;
        this.checked = true;
        this.color = "success";
    }
    hostData() {
        const { color } = this;
        return {
            class: Object.assign({}, createColorClasses(color), { [`checkbox-disabled`]: this.disabled, [`checkbox-disabled-checked`]: this.disabled && this.checked })
        };
    }
    handleCheck(event) {
        this.checked = event.target.checked;
        this.valueEmmit.emit(this.checked);
    }
    render() {
        return (h("div", { class: "checkbox" },
            h("input", { id: "checkbox__input", class: "checkbox__input", type: "checkbox", checked: this.checked, name: this.name, value: this.value, disabled: this.disabled, onChange: (event) => this.handleCheck(event) }),
            h("label", { htmlFor: "checkbox__input", class: "checkbox__label" }, this.checked &&
                (h("svg", { class: "icon-class", xmlns: "http://www.w3.org/2000/svg", width: "16", height: "12", viewBox: "0 0 16 12" },
                    h("path", { fill: "none", "fill-rule": "evenodd", "stroke-linecap": "round", "stroke-linejoin": "round", "stroke-width": "3", d: "M2 6l4 4 8-8" }))))));
    }
    static get is() { return "pulse-checkbox"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["pulse-checkbox.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["pulse-checkbox.css"]
    }; }
    static get properties() { return {
        "disabled": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "disabled",
            "reflect": false,
            "defaultValue": "false"
        },
        "name": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "name",
            "reflect": false
        },
        "value": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "value",
            "reflect": false
        },
        "checked": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "checked",
            "reflect": false,
            "defaultValue": "true"
        },
        "color": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "Color",
                "resolved": "\"error\" | \"pgray\" | \"primary\" | \"success\" | \"warn\"",
                "references": {
                    "Color": {
                        "location": "import",
                        "path": "../../../interface"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "color",
            "reflect": false,
            "defaultValue": "\"success\""
        }
    }; }
    static get events() { return [{
            "method": "valueEmmit",
            "name": "valueEmmit",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": ""
            },
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            }
        }]; }
}
