// @pulse.io/components/dist/collection/components/pulse-atm/tab/pulse-tab.js
import { h } from '@stencil/core';
import { createColorClasses } from '../../../utils/themes';
export class PulseTab {
    constructor() {
        this.inputId = `pulse-tab-${tabIds++}`;
        this.colorvariant = 'base';
        this.disabled = false;
        this.selected = false;
        this.onClick = () => {
            if (this.selected) {
                this.tabDeselect.emit();
            }
            else {
                this.selected = true;
            }
        };
    }
    selectChanged(isSelected) {
        if (isSelected) {
            this.tabSelect
                .emit({
                selected: true,
                value: this.value
            });
        }
    }
    componentDidLoad() {
        if (this.selected) {
            this.selectChanged(this.selected);
        }
        this.tabDidLoad.emit();
    }
    componentDidUnload() {
        this.tabDidUnload.emit();
    }
    componentWillLoad() {
        if (this.value === undefined) {
            this.value = this.inputId;
        }
    }
    hostData() {
        const { disabled, color, colorvariant } = this;
        return {
            class: Object.assign({}, createColorClasses(color, colorvariant), { 'button-disabled': disabled, 'pulse-focusable': true, 'tab-select--selected': this.selected })
        };
    }
    render() {
        return (h("div", { class: 'tab-select pulse-padding-xs-s-a', onClick: this.onClick },
            h("div", { class: "tab-select__inner" },
                h("div", { class: "tab-select__inner--content pulse-padding-xs-s-a" },
                    h("slot", null),
                    h("p", { class: 'tab-select__inner--content__tab-text' }, this.tabtext)))));
    }
    static get is() { return "pulse-tab"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["pulse-tab.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["pulse-tab.css"]
    }; }
    static get properties() { return {
        "color": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "Color",
                "resolved": "\"error\" | \"pgray\" | \"primary\" | \"success\" | \"warn\"",
                "references": {
                    "Color": {
                        "location": "import",
                        "path": "../../../interface"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "color",
            "reflect": false
        },
        "colorvariant": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "ColorVariant",
                "resolved": "\"base\" | \"dark\" | \"grad\" | \"shade\" | \"tint\"",
                "references": {
                    "ColorVariant": {
                        "location": "import",
                        "path": "../../../interface"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "colorvariant",
            "reflect": false,
            "defaultValue": "'base'"
        },
        "disabled": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "disabled",
            "reflect": true,
            "defaultValue": "false"
        },
        "tabtext": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "tabtext",
            "reflect": false
        },
        "value": {
            "type": "any",
            "mutable": true,
            "complexType": {
                "original": "any | null",
                "resolved": "any",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "value",
            "reflect": true
        },
        "selected": {
            "type": "boolean",
            "mutable": true,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "selected",
            "reflect": true,
            "defaultValue": "false"
        }
    }; }
    static get events() { return [{
            "method": "tabSelect",
            "name": "tabSelect",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Emitted when the select card is selected."
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }, {
            "method": "tabDeselect",
            "name": "tabDeselect",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [{
                        "text": undefined,
                        "name": "internal"
                    }],
                "text": "Emitted when selected select card is selected."
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }, {
            "method": "tabDidLoad",
            "name": "tabDidLoad",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [{
                        "text": undefined,
                        "name": "internal"
                    }],
                "text": "Emitted when the radio loads."
            },
            "complexType": {
                "original": "void",
                "resolved": "void",
                "references": {}
            }
        }, {
            "method": "tabDidUnload",
            "name": "tabDidUnload",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [{
                        "text": undefined,
                        "name": "internal"
                    }],
                "text": "Emitted when the radio unloads."
            },
            "complexType": {
                "original": "void",
                "resolved": "void",
                "references": {}
            }
        }]; }
    static get watchers() { return [{
            "propName": "selected",
            "methodName": "selectChanged"
        }]; }
}
let tabIds = 0;
