// @pulse.io/components/dist/collection/components/pulse-templates/flowtt/pulse-flowtt.js
/** * @author <a href="mailto:ivan.pena@avaldigitallabs.com">ivanntis</a> */
import { h } from '@stencil/core';
import { getBreakPointBySize } from '../../../utils/utils';
/**
 * Se define un  template que se adapta al 100% del ancho
 * del padre que lo contenga, y se cambian de posicion según
 * el tamaño del padre
 * Tiene 3 secciones contenedores:
 * 1 -> Cabecera
 * 2 -> Cuerpo a
 * 3 -> Cuerpo b
 */
export class PulseFlowtt {
    hostData() {
        return;
    }
    handleResize() {
        this.setTopSummaryCalc();
    }
    componentDidLoad() {
        this.setTopSummaryCalc();
    }
    setTopSummaryCalc() {
        const offsetBottom = (getBreakPointBySize(window.innerWidth) === 'lg') ? 24 : 0;
        this.summary.style.top = `${(this.head.offsetHeight + offsetBottom)}px`;
    }
    render() {
        return (h("div", { class: "flowtt-main" },
            h("div", { class: "flowtt-main--container pulse-grid" },
                h("header", { id: "header", class: "row flowtt-main--container--header", ref: el => this.head = el },
                    h("div", { class: "col-xs-12 col-md-10 col-lg-10 col-md-offset-1 col-lg-offset-1" },
                        h("slot", { name: "header" }))),
                h("section", { class: "row flowtt-main--container--content" },
                    h("div", { class: "col-xs-12 col-md-10 col-lg-3  col-md-offset-1 col-lg-offset-2 order-lg-2 flowtt-main--container--content--summary", ref: el => this.summary = el },
                        " ",
                        h("slot", { name: "summary" }),
                        " "),
                    h("div", { class: "col-xs-12  col-md-8 col-lg-5 col-md-offset-2 col-lg-offset-1 order-lg-1 flowtt-main--container--content--data" },
                        " ",
                        h("slot", { name: "content" }),
                        " ")))));
    }
    static get is() { return "pulse-flowtt"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["pulse-flowtt.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["pulse-flowtt.css"]
    }; }
    static get listeners() { return [{
            "name": "scroll",
            "method": "handleResize",
            "target": "window",
            "capture": false,
            "passive": true
        }]; }
}
