// @pulse.io/components/dist/collection/components/pulse-templates/flowtr/pulse-flowtr.js
/** * @author <a href="mailto:ivan.pena@avaldigitallabs.com">ivanntis</a> */
import { h } from '@stencil/core';
import { getBreakPointBySize } from '../../../utils/utils';
/**
 * Se define un  template que se adapta al 100% del ancho
 * del padre que lo contentenga. y se cambian de posicion segun
 * el tamaño del padre. esta pensado en flujos que sean maetsro
 * detalle y e cual el detalle tiene scroll y matestro siempre
 * esta fijo
 * Tiene 3 secciones contenedores:
 * 1 -> Cabecera
 * 2 -> Mestro
 * 3 -> Detalle
 */
export class PulseFlowtr {
    hostData() {
        return;
    }
    componentDidLoad() {
        this.setTopSummaryCalc();
    }
    setTopSummaryCalc() {
        const offsetBottom = (getBreakPointBySize(window.innerWidth) === 'lg') ? 24 : 0;
        this.master.style.top = `${(this.head.offsetHeight + offsetBottom)}px`;
    }
    render() {
        return (h("div", { class: "flowtr-main" },
            h("div", { class: "flowtr-main--container pulse-grid" },
                h("div", { class: "flowtr-main--container--bckg" },
                    h("header", { id: "header", class: "row flowtr-main--container--header", ref: el => this.head = el },
                        h("div", { class: "col-xs-12 col-md-10 col-lg-10 col-md-offset-1 col-lg-offset-1" },
                            h("slot", { name: "header" }))),
                    h("section", { id: "master", class: "row flowtr-main--container--master", ref: el => this.master = el },
                        h("div", { class: "col-xs-12 col-md-10 col-lg-10 col-md-offset-1 col-lg-offset-1" },
                            h("slot", { name: "master" }),
                            " "))),
                h("section", { id: "detail", class: "row flowtr-main--container--detail pulse-margin-xl-xs-v" },
                    h("div", { class: "col-xs-12 col-md-10  col-md-offset-1" },
                        " ",
                        h("slot", { name: "detail" }),
                        " ")))));
    }
    static get is() { return "pulse-flowtr"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["pulse-flowtr.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["pulse-flowtr.css"]
    }; }
}
