// @pulse.io/components/dist/collection/components/pulse-atm/fab-button/pulse-fab-button.js
import { h } from '@stencil/core';
import { hasShadowDom } from '../../../utils/utils';
import { createColorClasses } from '../../../utils/themes';
export class PulseFabButton {
    constructor() {
        this.disabled = false;
        this.type = 'button';
        this.textposition = 'bottom';
        this.onFocus = () => {
            this.pulseFocus.emit();
        };
        this.onBlur = () => {
            this.pulseBlur.emit();
        };
    }
    onClick(ev) {
        if (hasShadowDom(this.el)) {
            //Busca si hay un padre que contiene el elemento aislado por el shadowDom
            const form = this.el.closest('form');
            if (form) {
                ev.preventDefault();
                const virtualButton = document.createElement('button');
                virtualButton.type = this.type;
                virtualButton.style.display = 'none';
                form.appendChild(virtualButton);
                virtualButton.click();
                virtualButton.remove();
            }
        }
    }
    componentDidLoad() {
        this.fabButtonDidLoad.emit(this.el);
    }
    hostData() {
        const { disabled, color, textposition } = this;
        return {
            'aria-disabled': disabled ? 'true' : null,
            class: Object.assign({}, createColorClasses(color), { 'button': true, 'button-disabled': disabled, 'pulse-focusable': true, [`button-title-${textposition}`]: textposition })
        };
    }
    render() {
        const attrs = { type: this.type };
        return (h("div", { tabindex: "0", class: "button-main pulse-padding-s-xs-a" },
            h("button", Object.assign({ tabindex: "-1" }, attrs, { class: "button-native", onFocus: this.onFocus, onBlur: this.onBlur }),
                h("slot", null)),
            h("p", { tabindex: "-1", class: "button-title pulse-margin-s-xs-v" }, this.text)));
    }
    static get is() { return "pulse-fab-button"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["pulse-fab-button.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["pulse-fab-button.css"]
    }; }
    static get properties() { return {
        "text": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "text",
            "reflect": false
        },
        "color": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "Color",
                "resolved": "\"error\" | \"pgray\" | \"primary\" | \"success\" | \"warn\"",
                "references": {
                    "Color": {
                        "location": "import",
                        "path": "../../../interface"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "color",
            "reflect": false
        },
        "disabled": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "disabled",
            "reflect": true,
            "defaultValue": "false"
        },
        "type": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'submit' | 'reset' | 'button'",
                "resolved": "\"button\" | \"reset\" | \"submit\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "type",
            "reflect": false,
            "defaultValue": "'button'"
        },
        "textposition": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'bottom' | 'right'",
                "resolved": "\"bottom\" | \"right\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "textposition",
            "reflect": false,
            "defaultValue": "'bottom'"
        }
    }; }
    static get events() { return [{
            "method": "pulseFocus",
            "name": "pulseFocus",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": ""
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }, {
            "method": "pulseBlur",
            "name": "pulseBlur",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": ""
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }, {
            "method": "fabButtonDidLoad",
            "name": "fabButtonDidLoad",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": ""
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }]; }
    static get elementRef() { return "el"; }
    static get listeners() { return [{
            "name": "click",
            "method": "onClick",
            "target": undefined,
            "capture": false,
            "passive": false
        }]; }
}
