// @pulse.io/components/dist/collection/components/pulse-atm/card/pulse-card.js
/** * @author <a href="mailto:juan.gonzalez@avaldigitallabs.com">barto</a> */
import { h } from "@stencil/core";
import { createColorClasses } from '../../../utils/themes';
/**
 * Componente que genera un marco para tarjetas
 * incluye por defecto los colores basados en el tema
 * y las 6 elevaciones diseñadas en el Design System
 */
export class PulseCard {
    constructor() {
        this.elevation = 8;
        this.colorvariant = 'base';
        this.elevationhover = 0;
    }
    validateIndex(newValue) {
        const isValid = typeof newValue === 'number' && newValue % 2 === 0 && newValue <= 24;
        if (!isValid) {
            throw new Error('invalid elevation');
        }
        ;
    }
    hostData() {
        return {
            class: Object.assign({}, createColorClasses(this.color, this.colorvariant), { [`pulse-elevation-hover-${this.elevationhover}`]: !!this.elevationhover, [`pulse-elevation-${this.elevation}`]: true })
        };
    }
    render() {
        return (h("slot", null));
    }
    static get is() { return "pulse-card"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["pulse-card.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["pulse-card.css"]
    }; }
    static get properties() { return {
        "elevation": {
            "type": "number",
            "mutable": false,
            "complexType": {
                "original": "Elevation",
                "resolved": "0 | 16 | 2 | 24 | 4 | 8",
                "references": {
                    "Elevation": {
                        "location": "import",
                        "path": "../../../interface"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "elevation",
            "reflect": false,
            "defaultValue": "8"
        },
        "color": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "Color",
                "resolved": "\"error\" | \"pgray\" | \"primary\" | \"success\" | \"warn\"",
                "references": {
                    "Color": {
                        "location": "import",
                        "path": "../../../interface"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "color",
            "reflect": false
        },
        "colorvariant": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "ColorVariant",
                "resolved": "\"base\" | \"dark\" | \"grad\" | \"shade\" | \"tint\"",
                "references": {
                    "ColorVariant": {
                        "location": "import",
                        "path": "../../../interface"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "colorvariant",
            "reflect": false,
            "defaultValue": "'base'"
        },
        "elevationhover": {
            "type": "number",
            "mutable": false,
            "complexType": {
                "original": "Elevation",
                "resolved": "0 | 16 | 2 | 24 | 4 | 8",
                "references": {
                    "Elevation": {
                        "location": "import",
                        "path": "../../../interface"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "elevationhover",
            "reflect": false,
            "defaultValue": "0"
        }
    }; }
    static get watchers() { return [{
            "propName": "elevation",
            "methodName": "validateIndex"
        }]; }
}
